ibmmq/mqicb.go

```diff
@@ -0,0 +1,210 @@
+
+		gocbc.CallbackArea = info.callbackArea
+		gocbc.ConnectionArea = info.connectionArea
+
+		// Get the data
+		b := C.GoBytes(unsafe.Pointer(mqBuffer), C.int(mqcbc.DataLength))
+
+		// And finally call the user function
+		info.callbackFunction(info.hObj, gomd, gogmo, b, gocbc, mqreturn)
+	}
+}
+
+/*
+CB is the function to register/unregister a callback function for a queue, based on
+criteria in the message descriptor and get-message-options
+*/
+func (object *MQObject) CB(goOperation int32, gocbd *MQCBD, gomd *MQMD, gogmo *MQGMO) error {
+	var mqrc C.MQLONG
+	var mqcc C.MQLONG
+	var mqOperation C.MQLONG
+	var mqcbd C.MQCBD
+	var mqmd C.MQMD
+	var mqgmo C.MQGMO
+
+	mqOperation = C.MQLONG(goOperation)
+	copyCBDtoC(&mqcbd, gocbd)
+	copyMDtoC(&mqmd, gomd)
+	copyGMOtoC(&mqgmo, gogmo)
+
+	key := makeKey(object.qMgr.hConn, object.hObj)
+
+	// The callback function is a C function that is a proxy for the MQCALLBACK_Go function
+	// defined here. And that in turn will call the user's callback function
+	mqcbd.CallbackFunction = (C.MQPTR)(unsafe.Pointer(C.MQCALLBACK_C))
+
+	C.MQCB(object.qMgr.hConn, mqOperation, (C.PMQVOID)(unsafe.Pointer(&mqcbd)),
+		object.hObj,
+		(C.PMQVOID)(unsafe.Pointer(&mqmd)), (C.PMQVOID)(unsafe.Pointer(&mqgmo)),
+		&mqcc, &mqrc)
+
+	mqreturn := MQReturn{MQCC: int32(mqcc),
+		MQRC: int32(mqrc),
+		verb: "MQCB",
+	}
+
+	if mqcc != C.MQCC_OK {
+		return &mqreturn
+	}
+
+	// Add or remove the control information in the map used by the callback routines
+	switch mqOperation {
+	case C.MQOP_DEREGISTER:
+		delete(cbMap, key)
+	case C.MQOP_REGISTER:
+		// Stash the hObj and real function to be called
+		info := &cbInfo{hObj: object,
+			callbackFunction: gocbd.CallbackFunction,
+			connectionArea:   nil,
+			callbackArea:     gocbd.CallbackArea}
+		cbMap[key] = info
+	default: // Other values leave the map alone
+	}
+
+	return nil
+}
+
+/*
+Ctl is the function that starts/stops invocation of a registered callback.
+*/
+func (x *MQQueueManager) Ctl(goOperation int32, goctlo *MQCTLO) error {
+	var mqrc C.MQLONG
+	var mqcc C.MQLONG
+	var mqOperation C.MQLONG
+	var mqctlo C.MQCTLO
+
+	mqOperation = C.MQLONG(goOperation)
+	copyCTLOtoC(&mqctlo, goctlo)
+
+	// Need to make sure control information is available before the callback
+	// is enabled. So this gets setup even if the MQCTL fails.
+	key := makePartialKey(x.hConn)
+	for k, info := range cbMap {
+		if strings.HasPrefix(k, key) {
+			info.connectionArea = goctlo.ConnectionArea
+		}
+	}
+
+	C.MQCTL(x.hConn, mqOperation, (C.PMQVOID)(unsafe.Pointer(&mqctlo)), &mqcc, &mqrc)
+
+	mqreturn := MQReturn{MQCC: int32(mqcc),
+		MQRC: int32(mqrc),
+		verb: "MQCTL",
+	}
+
+	if mqcc != C.MQCC_OK {
+		return &mqreturn
+	}
+
+	return nil
+}
+
+// Functions below here manage the map of objects and control information so that
+// the Go variables can be saved/restored from invocations to the C layer
+func makeKey(hConn C.MQHCONN, hObj C.MQHOBJ) string {
+	key := fmt.Sprintf("%d/%d", hConn, hObj)
```

Comment:
Could just do: `return fmt.Sprintf("%d/%d", hConn, hObj)` instead of setting variable to return it.